### PR TITLE
GradNorm adaptive loss weighting for tandem-transfer

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Fixed PCGrad loss weights treat all tasks (in_dist, ood_cond, tandem_transfer, ood_re) as equally learnable, but tandem_transfer (p_tan) consistently dominates the error. GradNorm (Chen et al., 2018) dynamically adjusts per-task loss weights so that all gradient norms grow at the same normalized rate relative to the initial step — preventing easy tasks from crowding out hard ones.

The hypothesis is that p_tan has a smaller gradient magnitude on the shared backbone than in_dist tasks (because tandem configurations are rarer and harder), causing the optimizer to under-invest in it. GradNorm will detect this imbalance and up-weight tandem losses automatically, with the asymmetry parameter alpha=1.5 biasing towards tasks that are lagging (i.e. tandem).

Reference: "GradNorm: Gradient Normalization for Adaptive Loss Balancing in Deep Multitask Networks", Chen et al. 2018, ICML. https://arxiv.org/abs/1711.02257

## Instructions

Add a `--gradnorm` flag and `--gradnorm_alpha` float flag (default 1.5) to `train.py`. The implementation:

1. **Add GradNorm task weights**: Add a `nn.Parameter` vector `task_weights = nn.Parameter(torch.ones(N_tasks))` where N_tasks is the number of loss components being balanced (in_dist, ood_cond, tandem_transfer, ood_re single-foil each).

2. **Track initial loss values**: On the first training step, record `L0_i` (the initial loss per task) as a non-learnable buffer. These are the baseline values GradNorm normalises against.

3. **GradNorm weight update** (runs *after* the regular backward pass, before optimizer.step()):
   - Compute the gradient norm `G_i` of each task loss w.r.t. the *last shared layer* parameters (the final TransolverBlock output projection — the `fc2` linear in the last block). Use `torch.autograd.grad(..., retain_graph=True)` with `only_inputs=True`.
   - Compute mean gradient norm: `G_mean = mean(G_i * task_weights_i)` — note task_weight-weighted mean.
   - Compute loss ratio: `r_i = (L_i / L0_i) / mean(L_j / L0_j)` — normalized relative inverse training rate.
   - GradNorm target: `G_target_i = G_mean * r_i ** alpha` (detached).
   - GradNorm loss: `L_grad = sum(|G_i * task_weights_i - G_target_i|)` — L1 norm of gradient mismatch.
   - Run a **separate** `L_grad.backward()` on `task_weights` only (zero grad on model params first). Update `task_weights` with a separate Adam optimizer, lr=1e-3.
   - After update, renormalize: `task_weights.data = task_weights.data / task_weights.data.mean() * N_tasks` — keeps the total loss scale stable.

4. **Apply task weights to the main loss**: Multiply each task's loss component by `task_weights[i].detach()` before summing into the total loss for the main backward pass.

5. **Use EMA smoothing** for the task weight updates: rather than directly using GradNorm's new weight, apply `w_new = 0.9 * w_old + 0.1 * w_gradnorm` to avoid rapid oscillation.

6. **Log task weights** to W&B each epoch: `wandb.log({"task_weight/in_dist": ..., "task_weight/tandem": ..., ...})`.

The PCGrad gradient surgery (`--pcgrad_3way`) should remain active and is compatible — GradNorm operates on task weights (scalar multipliers on the loss), while PCGrad operates on gradient directions. They address orthogonal aspects.

Run with these flags (add to the baseline command):
```
--gradnorm --gradnorm_alpha 1.5
```

Run **two seeds**: `--seed 42` and `--seed 73`. Use `--wandb_group gradnorm-alpha1.5` for both runs.

Full training command (seed 42):
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/gradnorm-alpha1.5-s42" --wandb_group gradnorm-alpha1.5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --gradnorm --gradnorm_alpha 1.5
```

## Baseline

Current best metrics (PR #2184 — DCT Freq Loss w=0.05, 2-seed avg):

| Metric | Baseline |
|--------|----------|
| p_in   | 13.205   |
| p_oodc | 7.816    |
| p_tan  | 28.502   |
| p_re   | 6.453    |

## Results

_To be filled in by thorfinn after training runs complete._

| Run | Seed | p_in | p_oodc | p_tan | p_re | W&B run |
|-----|------|------|--------|-------|------|---------|
| gradnorm-alpha1.5 | 42 | — | — | — | — | — |
| gradnorm-alpha1.5 | 73 | — | — | — | — | — |
| **2-seed avg** | — | — | — | — | — | — |
| **vs baseline** | — | — | — | — | — | — |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### GradNorm Adaptive Loss Weighting (alpha=1.5)

| Run | Seed | p_in | p_oodc | p_tan | p_re | W&B run |
|-----|------|------|--------|-------|------|---------|
| gradnorm-alpha1.5 | 42 | 12.9 | 7.6 | 28.6 | 6.4 | `3wkhlz4a` |
| gradnorm-alpha1.5 | 73 | 13.2 | 8.0 | 29.3 | 6.5 | `5aw7ac8m` |
| **2-seed avg** | — | **13.05** | **7.80** | **28.95** | **6.45** | — |
| **Baseline** | — | 13.205 | 7.816 | 28.502 | 6.453 | `6yfv5lio`/`etepxvjc` |
| **Δ vs baseline** | — | **−1.2%** | **−0.2%** | **+1.6%** | **−0.05%** | — |

### Commands used

```bash
# Seed 42 (GPU 0)
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent thorfinn --seed 42 \
  --wandb_name "thorfinn/gradnorm-alpha1.5-s42" --wandb_group "gradnorm-alpha1.5" \
  --gradnorm --gradnorm_alpha 1.5 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73 (GPU 1) — same flags with --seed 73
```

### Training details

- **S42**: Best at epoch 149/149, val/loss=0.3821, 180.2 min, 44.8GB peak VRAM
- **S73**: Best at epoch 148/148, val/loss=0.3910, 180.9 min, 46.2GB peak VRAM
- GradNorm reference layer: last TransolverBlock's MLP output projection (73,920 params)
- GradNorm tasks: 3-way (single-foil | tandem-normal | tandem-extreme), matching PCGrad grouping
- Separate Adam optimizer for task weights (lr=1e-3), EMA smoothing (0.9/0.1)

### What happened

**Essentially neutral result.** GradNorm with alpha=1.5 produced metrics within noise of the baseline across all splits. The small p_in improvement (−1.2%) is offset by a slight p_tan regression (+1.6%). OOD metrics (p_oodc, p_re) are virtually unchanged.

**Notable behavior — S73 delayed convergence:** Seed 73 exhibited a striking convergence delay where val_in_dist was stuck at ~15 (10× worse than S42) for the first 11 epochs, then snapped to ~1.5 at epoch 12 and tracked normally thereafter. This suggests GradNorm's task weight initialization can create a period of instability before the weights stabilize, and the outcome is seed-dependent. S42 converged smoothly from epoch 1.

**Why it didn't help:** The hypothesis was that tandem_transfer has smaller gradient magnitude, causing the optimizer to under-invest in it. GradNorm was expected to detect and correct this. However, the neutral result suggests that:
1. **PCGrad 3-way already addresses the gradient imbalance** — it projects conflicting gradients, which may handle the same problem GradNorm targets but at the direction level rather than magnitude level.
2. **The tandem_transfer bottleneck isn't gradient magnitude** — it may be a data/representation limitation (tandem configurations are physically harder to predict due to wake interactions, not just underweighted).
3. **GradNorm + PCGrad may partially compete** — both try to rebalance task learning rates, and layering them produces no net benefit.

### Suggested follow-ups

1. **GradNorm without PCGrad** — test GradNorm alone (disable `--pcgrad_3way`) to see if they compete or complement. If GradNorm alone underperforms, it confirms PCGrad is the more effective mechanism.
2. **Lower alpha (0.5)** — less aggressive rebalancing may avoid the S73-style delayed convergence and could be more stable.
3. **Task-specific learning rates** — instead of scalar loss weights, use per-task learning rate multipliers on the optimizer, which affects parameter updates more directly.
4. **Tandem-specific data augmentation** — if the bottleneck is representation quality rather than gradient balance, augmenting tandem configurations specifically may be more effective.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2198 because Dead end. GradNorm adaptive loss weighting regressed p_tan from 28.502→28.979 (+1.7%), the exact metric it was designed to improve. The hypothesis was that tandem tasks are under-weighted due to smaller gradient magnitudes, but GradNorm's dynamic rebalancing actually disrupted the well-tuned PCGrad gradient surgery balance — PCGrad and GradNorm are not fully orthogonal in practice when combined with EMA smoothing. p_in improved slightly (13.205→13.070, -1.0%) but that's insufficient to offset the p_tan regression.

W&B run IDs: 3wkhlz4a (seed 42, p_tan=28.611), 5aw7ac8m (seed 73, p_tan=29.347). 2-seed avg: p_tan=28.979 vs baseline 28.502..
